lib/swoosh/adapters/gmail.ex

```diff
@@ -0,0 +1,145 @@
+  @moduledoc """
+  An adapter that sends email using Gmail api
+
+  For reference [Gmail API docs](https://developers.google.com/gmail/api)
+
+  You don't need to set `from` address as google will set it for you.
+  If you still want to include it, make sure it matches the account or
+  it will be ignored.
+
+  ## Dependency
+
+  Gmail adapter requires `Mail` dependency to format message as RFC 2822 message.
+
+  Because `Mail` library removes Bcc headers, they are being added after email is
+  rendered.
+  ## Example
+
+      # config/congig.exs
+      config :sample, Smaple.Mailer
+        adapter: Swoosh.Adapters.Gmail,
+
+
+      # lib/sample/mailer.ex
+      defmodule Sample.Mailer do
+        use Swoosh.Mailer, otp_app: :sample
+      end
+
+  ## Required config parameters
+    - `:access_token` valid OAuth2 access token
+        Required scopes:
+        - gmail.compose
+      See https://developers.google.com/oauthplayground when developing
+  """
+
+  use Swoosh.Adapter, required_deps: [mail: Mail]
+
+  alias Swoosh.Email
+
+  @base_url "https://www.googleapis.com/upload/gmail/v1"
+  @api_endpoint "/users/me/messages/send"
+
+  def deliver(%Email{} = email, config \\ []) do
+    access_token = config[:access_token] || raise(ArgumentError, "access_token is required")
+
+    headers = [
+      {"Authorization", "Bearer #{access_token}"},
+      {"Content-Type", "message/rfc822"}
+    ]
+
+    url = [base_url(config), @api_endpoint]
+
+    body = prepare_body(email)
+
+    case Swoosh.ApiClient.post(url, headers, body, email) do
+      {:ok, 200, _headers, body} ->
+        {:ok, parse_response(body)}
+
+      {:ok, code, _headers, body} when code >= 400 and code <= 599 ->
+        {:error, {code, parse_response(body)}}
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  defp parse_response(body) when is_binary(body),
+    do: body |> Swoosh.json_library().decode! |> parse_response()
+
+  defp parse_response(%{"id" => id, "threadId" => thread_id, "labelIds" => labels}) do
+    %{id: id, thread_id: thread_id, labels: labels}
+  end
+
+  defp parse_response(%{"error" => %{"errors" => errors, "code" => code, "message" => message}}) do
+    %{error: %{code: code, message: message}, errors: Enum.map(errors, &parse_error/1)}
+  end
+
+  defp parse_error(error) do
+    %{
+      domain: error["domain"],
+      reason: error["reason"],
+      message: error["message"],
+      location_type: error["locationType"],
+      location: error["location"]
+    }
+  end
+
+  defp base_url(config), do: config[:base_url] || @base_url
+
+  def prepare_body(email) do
+    Mail.build_multipart()
+    |> prepare_from(email)
+    |> prepare_to(email)
+    |> prepare_cc(email)
+    |> prepare_bcc(email)
+    |> prepare_subject(email)
+    |> prepare_text(email)
+    |> prepare_html(email)
+    |> prepare_attachments(email)
+    |> prepare_reply_to(email)
+    |> Mail.Renderers.RFC2822.render()
+    |> parse_bcc(email)
```

Comment:
worth noting(maybe should document also) that this is on purpose, because `Mail` will remove any `bcc` header when rendering, so we have to put it back after to be able to use the functionality. Wasn't able to dig out the exact reason for it, but if anyone has any info on that would love to know.

Reply:
Yep, a comment will be super helpful in the future, in case my OCD kicks in. Also this function should probably be private.

Reply:
Ah good catch. It was private, but changed it to access directly while doing some testing. Will update and add the comment